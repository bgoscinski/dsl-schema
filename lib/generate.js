"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const J = require("./json-types");
const T = require("./schema-types");
const register_1 = require("./register");
const DATE_TIME_LIKE = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z$/;
const URL_LIKE = /^[a-z][a-z0-9+.-]*:/;
const EMAIL_LIKE = /^.{1,30}@.{1,20}\..{1,4}$/;
const IPV4_LIKE = /^(?:[0-9]{1,3}\.){3}[0-9]{1,3}$/;
const IPV6_LIKE = /^(?:[0-9a-f]{0,4}:){2,7}(?:[0-9a-f]{1,4}$)$/i;
const UUID_LIKE = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i;
function sthLike(example) {
    if (register_1.isRegistered(example))
        return example;
    if (J.isNullLike(example))
        return T.NULL;
    if (J.isBooleanLike(example))
        return T.BOOL;
    if (J.isNumberLike(example)) {
        if (Number.isInteger(example))
            return T.Int();
        return T.Float();
    }
    if (J.isStringLike(example)) {
        if (DATE_TIME_LIKE.test(example))
            return T.DATE_TIME;
        if (URL_LIKE.test(example))
            return T.URL;
        if (EMAIL_LIKE.test(example))
            return T.EMAIL;
        if (IPV4_LIKE.test(example))
            return T.IPV4;
        if (IPV6_LIKE.test(example))
            return T.IPV6;
        if (UUID_LIKE.test(example))
            return T.UUID;
        return T.Str();
    }
    if (J.isArrayLike(example)) {
        return T.Tuple(Array.from(example)
            .map(sthLike)
            .map(T.req));
    }
    if (J.isObjectLike(example)) {
        const props = Object.keys(example).reduce((acc, name) => {
            acc[name] = T.req(sthLike(example[name]));
            return acc;
        }, {});
        return T.Record(props);
    }
    throw TypeError(`${sthLike.name}(example): don't know how to create schema for ${example}`);
}
exports.sthLike = sthLike;
