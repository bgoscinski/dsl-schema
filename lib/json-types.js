"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const { toString } = Object.prototype;
function isNullLike(x) {
    return x == null;
}
exports.isNullLike = isNullLike;
function isBooleanLike(x) {
    return x === true || x === false || toString.call(x) === '[object Boolean]';
}
exports.isBooleanLike = isBooleanLike;
function isNumberLike(x) {
    return typeof x === 'number' || toString.call(x) === '[object Number]';
}
exports.isNumberLike = isNumberLike;
function isStringLike(x) {
    return typeof x === 'string' || toString.call(x) === '[object String]';
}
exports.isStringLike = isStringLike;
function isArrayLike(x) {
    return (x != null &&
        (Array.isArray(x) ||
            (typeof x === 'object' && isNumberLike(x.length) && !isStringLike(x))));
}
exports.isArrayLike = isArrayLike;
function isObjectLike(x) {
    return x != null && toString.call(x) === '[object Object]';
}
exports.isObjectLike = isObjectLike;
