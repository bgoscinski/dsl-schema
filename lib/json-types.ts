const { toString } = Object.prototype;

export function isNullLike(x) {
  return x == null;
}

export function isBooleanLike(x) {
  return x === true || x === false || toString.call(x) === '[object Boolean]';
}

export function isNumberLike(x) {
  return typeof x === 'number' || toString.call(x) === '[object Number]';
}

export function isStringLike(x) {
  return typeof x === 'string' || toString.call(x) === '[object String]';
}

export function isArrayLike(x) {
  return (
    x != null &&
    (Array.isArray(x) ||
      (typeof x === 'object' && isNumberLike(x.length) && !isStringLike(x)))
  );
}

export function isObjectLike(x) {
  return x != null && toString.call(x) === '[object Object]';
}
