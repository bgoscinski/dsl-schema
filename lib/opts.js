export const isOptsStr = x => typeof x === 'string';

function tryParseUniq(str) {
  if (str === 'uniq' || str === 'unique') {
    return { type: 'uniq' };
  }
}

export const uniq = () => opt => opt.type === 'uniq';

const getLeftBoundary = ([op, incl], val) => {
  const ret = op === '<' ? { min: val } : { max: val };
  if (!incl) ret[op === '<' ? 'exclusiveMin' : 'exclusiveMax'] = true;
  return ret;
};

const getRightBoundary = ([op, incl], val) => {
  const ret = op === '>' ? { min: val } : { max: val };
  if (!incl) ret[op === '>' ? 'exclusiveMin' : 'exclusiveMax'] = true;
  return ret;
};

//                       | lhs   |   | lop  |     | bind |      | rop  |   | rhs   |
const RANGE_REGEX = /^(?:([\d\.]+)\s*([<>]=?)\s*)?([a-z]+)(?:\s*([<>]=?)\s*([\d\.]+))?$/i;

function tryParseRangeOpt(str) {
  const match = str.trim().match(RANGE_REGEX);

  if (!match) return;
  const [, lhs, lop, binding, rop, rhs] = match;
  if (!lop && !rop) return;
  if (lop && rop && lop.charAt(0) !== rop.charAt(0)) return;

  return {
    type: 'range',
    binding,
    ...(lop && getLeftBoundary(lop, parseFloat(lhs))),
    ...(rop && getRightBoundary(rop, parseFloat(rhs))),
  };
}

export const range = binding => opt =>
  opt.type === 'range' && opt.binding === binding;

//                      | lhs   |          | bind |           | rhs   |
const MULT_REGEX = /^(?:([\d\.]+)\s*\*\s*)?([a-z]+)(?:\s*\*\s*([\d\.]+))?$/i;

function tryParseMult(str) {
  const match = str.trim().match(MULT_REGEX);

  if (!match) return;
  const [, lhs, binding, rhs] = match;
  if (lhs && rhs) return;
  if (!lhs && !rhs) return;

  return { type: 'mult', binding, multiplier: parseFloat(lhs || rhs) };
}

export const mult = binding => opt =>
  opt.type === 'mult' && opt.binding === binding;

export function minIntFromExclusiveFloat(exclusiveMin) {
  if (Number.isInteger(exclusiveMin)) {
    return exclusiveMin + 1;
  }
  return Math.ceil(exclusiveMin);
}

export function maxIntFromExclusiveFloat(exclusiveMax) {
  if (Number.isInteger(exclusiveMax)) {
    return exclusiveMax - 1;
  }
  return Math.floor(exclusiveMax);
}

export function minIntFromInclusiveFloat(inclusiveMin) {
  return Math.ceil(inclusiveMin);
}

export function maxIntFromInclusiveFloat(inclusiveMax) {
  return Math.floor(inclusiveMax);
}

export function parse(spec) {
  if (!spec) return [];
  return spec.split(/\s*,\s*/).reduce((acc, part) => {
    const parsed =
      tryParseUniq(part) || tryParseRangeOpt(part) || tryParseMult(part);
    return parsed ? [...acc, parsed] : acc;
  }, []);
}
