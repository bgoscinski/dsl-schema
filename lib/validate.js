import Ajv from 'ajv';
const ajv = new Ajv({ allErrors: true });

export function validatorFor(schema) {
  const validate = ajv.compile(schema);
  return function schemaValidator(data) {
    const isValid = validate(data);
    const errors = validate.errors;
    validate.errors = null;
    return { isValid, errors };
  };
}
