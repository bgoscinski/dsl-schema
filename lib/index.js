import * as T from './schema-types';
import * as G from './generate';
import { wrapFactory, regAsSchema } from './register';

export { validatorFor } from './validate';

export const sthLike = wrapFactory(G.sthLike);

export const str = wrapFactory(T.str);
export const int = wrapFactory(T.int);
export const float = wrapFactory(T.float);
export const record = wrapFactory(T.record);
export const dict = wrapFactory(T.dict);
export const list = wrapFactory(T.list);
export const tuple = wrapFactory(T.tuple);
export const allOf = wrapFactory(T.allOf);
export const anyOf = wrapFactory(T.anyOf);
export const oneOf = wrapFactory(T.oneOf);
export const not = wrapFactory(T.not);

export const Bool = regAsSchema(T.Bool);
export const Null = regAsSchema(T.Null);
