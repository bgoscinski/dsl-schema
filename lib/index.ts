import * as Public from './public-api';
import { sthLike } from './generate';

export * from './public-api';

export default Object.assign(function like(example) {
  return sthLike(example);
}, Public);
