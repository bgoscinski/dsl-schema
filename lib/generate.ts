import * as J from './json-types'
import * as T from './schema-types'
import { isRegistered, isTagged } from './register'

const DATE_TIME_LIKE = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z$/
const URL_LIKE = /^[a-z][a-z0-9+.-]*:/
const EMAIL_LIKE = /^.{1,30}@.{1,20}\..{1,4}$/
const IPV4_LIKE = /^(?:[0-9]{1,3}\.){3}[0-9]{1,3}$/
const IPV6_LIKE = /^(?:[0-9a-f]{0,4}:){2,7}(?:[0-9a-f]{1,4}$)$/i
const UUID_LIKE = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i

export function sthLike(example: unknown): T.Schema {
  if (isRegistered(example)) return example

  if (J.isNullLike(example)) return T.NULL
  if (J.isBooleanLike(example)) return T.BOOL
  if (J.isNumberLike(example)) {
    if (Number.isInteger(+example)) return T.Int()
    return T.Float()
  }
  if (J.isStringLike(example)) {
    if (DATE_TIME_LIKE.test('' + example)) return T.DATE_TIME
    if (URL_LIKE.test('' + example)) return T.URL
    if (EMAIL_LIKE.test('' + example)) return T.EMAIL
    if (IPV4_LIKE.test('' + example)) return T.IPV4
    if (IPV6_LIKE.test('' + example)) return T.IPV6
    if (UUID_LIKE.test('' + example)) return T.UUID
    return T.Str()
  }
  if (J.isObjectLike(example)) {
    const props = Object.keys(example).reduce<{ [k: string]: T.MemberTag }>(
      (acc, name) => {
        const val = example[name]
        acc[name] = isTagged(val) ? val : T.req(sthLike(val))
        return acc
      },
      {}
    )

    return T.Record(props)
  }
  if (J.isArrayLike(example)) {
    return T.Tuple(
      Array.from(example).map(ex => {
        return isTagged(ex) ? ex : T.req(sthLike(ex))
      })
    )
  }

  throw TypeError(
    `${sthLike.name}(example): don't know how to create schema for ${example}`
  )
}
