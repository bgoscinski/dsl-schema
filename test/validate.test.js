import { assertFor } from '../lib/validate';
import { sthLike } from '../lib/generate';

describe(assertFor.name, () => {
  const assert = assertFor(
    sthLike({
      a: 1,
      b: {
        c: 3,
        d: [4, 5],
      },
    })
  );

  it(`should not throw when data matches schema`, () => {
    expect(() =>
      assert({
        a: 6,
        b: {
          c: 7,
          d: [8, 9],
        },
      })
    ).not.toThrow();
  });

  it(`should throw when data doesn't match schema [1]`, () => {
    expect(() => assert()).toThrowErrorMatchingSnapshot();
  });

  it(`should throw when data doesn't match schema [2]`, () => {
    expect(() =>
      assert({ a: '', b: { d: [''] } })
    ).toThrowErrorMatchingSnapshot();
  });
});
