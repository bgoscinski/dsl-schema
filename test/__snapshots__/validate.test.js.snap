// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assertFor should throw when data doesn't match schema [1] 1`] = `
"[1] Checked value doesn't match schema
  Reason: should be object
  Value:
    [90mundefined[39m
  Schema:
    [32m'object'[39m
"
`;

exports[`assertFor should throw when data doesn't match schema [2] 1`] = `
"[1] Property at .a doesn't match schema
  Reason: should be number
  Value:
    [32m''[39m
  Schema:
    [32m'number'[39m

[2] Property at .b doesn't match schema
  Reason: should have required property 'c'
  Value:
    { d: [36m[Array][39m }
  Schema:
    { c: [36m[Object][39m, d: [36m[Object][39m }

[3] Property at .b.d[0] doesn't match schema
  Reason: should be number
  Value:
    [32m''[39m
  Schema:
    [32m'number'[39m
"
`;
